Kernel_23/include/CGAL/Kernel_checker.h

```diff
@@ -80,19 +79,16 @@ class Primitive_checker
   { }
 
   template <class ... A>
-  typename Pairify<typename CGAL::cpp11::result_of<P1(const A&...)>::type,
-                   typename CGAL::cpp11::result_of<P2(const A&...)>::type>::result_type
+  decltype(auto)
   operator()(const A&... a) const
   {
-    typedef typename CGAL::cpp11::result_of<P1(const A&...)>::type result_type_1;
-    typedef typename CGAL::cpp11::result_of<P2(const A&...)>::type result_type_2;
-    result_type_1 res1 = p1(a.first...);
-    result_type_2 res2 = p2(a.second...);
+    auto res1 = p1(a.first...);
+    auto res2 = p2(a.second...);
```

Comment:
I actually did not check that. I will!

Reply:
No, they are never reference. I have also tried to use `auto&&` and it actually breaks the code. So I left it as it is.

Reply:
@mglisse Why `auto&&`? would not `decltype(auto)` be better in that case?

Reply:
As far as I understood, this simply returns an object that is always returned by value, so `auto` is enough IMHO. It simply catches the object's type.

Reply:
The argument is `CGAL::Point_3<CGAL::Epick>`, and `p1` is `CGAL::Epick::Compute_x_3()`, then I am pretty sure the result should be `const double&`.

Reply:
Ok, I will check that.

Reply:
> Why `auto&&`? would not `decltype(auto)` be better in that case?

Because that's the usual way to use them? `auto&&` when you declare a variable (and in particular a function argument), and decltype(auto) for the return type of a function. It might make sense to use decltype(auto) when declaring a variable, but from what I remember, the few times we thought we saw a case for this, it turned out that auto&& was a better choice.

Reply:
I carried on more checks using `boost::typeindex` and, indeed, sometimes `p1` and `p2` return const ref, however, I compared the results of `auto` with `result_type_1` and `result_type_2` and they are identical while when using `auto&&` or `decltype(auto)` or I also tried a few other ways to write this method, none of them either compiles or works or gives anything close to what it was before. Moreover, both `res1` and `res2` are used only in `Pairify` where, as far as I see, `auto` is enough.